src/java.base/share/classes/java/net/URL.java

```diff
@@ -1092,7 +1092,13 @@ public URI toURI() throws URISyntaxException {
      *             int, java.lang.String)
      */
     public URLConnection openConnection() throws java.io.IOException {
-        return handler.openConnection(this);
+        URLConnection url = null;
+        try {
+            url = handler.openConnection(this);
+        } catch (IllegalArgumentException e) {
+            throw new MalformedURLException(e.getMessage());
+        }
+        return url;
```

Comment:
Please revert the changes in java.net.URL. It's not the proper place to do the conversion. If a conversion is needed, it should take place in the handler.
One possible experiment could be to  change `ParseUtil::decode` to declare `throws MalformedURLException` and see what no longer compiles. If compilation passes - then maybe that's the way to go. If it doesn't - then we could analyze what to do on a case-by-case basis.

Reply:
There are some methods call ParseUtil::decode and of these 10 are not declared to be thrown MalformedURLException.
Modifying ParseUtil :: decode has not a little impact. Is modifying java.net.URL not proper?

The following classes have the methods which are not declared to be thrown MalformedURLException.

```
java.net.URLClassLoader
sun.net.www.protocol.ftp.FtpURLConnection
sun.security.provider.PolicyFile
jdk.internal.loader.FileURLMapper
```

Reply:
@dfuch 
Could you tell me your opinion based on this information, please?

Reply:
Let me try to summarize: there are several things that we could do:

 1. nothing: continue throwing IIOBE
 2. throw IAE instead of IIOBE, and let IAE trickle up the calling stack to the callers code
 3. throw MUE instead of IIOBE, and then update all the places where ParseUtils::decode is called to either let the MUE trickle up the stack if possible, or deal with it if not.
 4. throw IAE instead of IIOBE and somewhere up the stack catch that and transform it into MUE.
   4.1. do the catch & transform in URL::openConnection when calling URLStreamHandler::openConnection
   4.2. do the catch & transform in every protocol handler implementation that use ParseUtils::decode
    
Now let's see pro & cons:
 1. is not ideal because it looks like a bug (and it is)
 2. is expedient. There seem to be some precedent for that since ParseUtils::decode already throws IAE if it can't parse the port. However IAE is not specified to be thrown by openConnection. So maybe we should update the specification of URL::openConnection and URLStreamHandler::openConnection to allow for IAE. This is however a bit strange to use IAE because in this case the URL is indeed malformed, so MUE would be more appropriate.
 3. has the advantage of being clear, and force all callers of ParseUtils to deal with the situation. However, ParseUtils seems to be widely used, so each places where it's called needs to be examined, to see what would be the appropriate solution in each case 
 4. The advantage here is that callers of ParseUtils don't need to be changed if IAE is already appropriate for them. But those for which MUE would be more appropriate would beneficiate from having IAE transformed in MUE, which can be achieved by 4.1 or 4.2
  4.1 the issue here is that URLStreamHandler::openConnection is not specified to throw IAE, yet we catch IAE that might be thrown by custom URLStreamHandler implementation not present in the JDK, where the IAE that they throw might have nothing to do with a Malformed URL. So we risk introducing regressions for applications that provide their own custom URLStreamHandlers implementations, and that might not expect the wrapping of the IAE they throw into MUE.  I believe that if we did that it would need to be specified (in URL::openConnection) and it would require a CSR.
  4.2 we need to examine all protocol handler present in the JDK and do the wrapping there. URLStreamHandler::openConnection is specified to throw IOException, so throwing MUE is within the bounds. However - if we don't have ParseUtils::decode throw MUE, we're not guaranteed not to miss some places.
  
I'd say from a consistency point of view, 3. is probably best. Then failing that, from a compatibility vs consistency point of view 4.2 might be our best shot. 2. might be envisaged, but maybe at the expense of logging a follow up issue to clean that up and later decide between 3. or 4.2.

Reply:
Thank you for your valuable comments. I think 3 is a good idea.  I will try fix, but it takes some time to see the impact on many other methods.

Reply:
@dfuch 

I'm sorry, it took time for me to think about how to fix.
First, I thought of option 3, but it seems difficult.
In many of the methods in error, the javadoc defines the exceptions that can be thrown. 
When option 3, we will wrap exceptions in many methods because it is difficult to fix specifications of all methods and CSR.
However, if we are going to wrap exceptions, I think it's better and less fix to do it in a protocol handler, it means option 4.2.

Let me hear what you think.

Reply:
4.2 might be fine especially if we can do it at the point were ParseUtils is called.

Reply:
@dfuch 
Thank you very much. Your advice was very helpful in fix.
I fixed sun.net.www.protocol.ftp.Hanler.openConnection to throw MalformedURLException. 
Hanler classes in other protcol should not be fixed because they don't use ParseUtil.decode method.